torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp

```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
   return AliasAnalysisKind::FROM_SCHEMA;
 }
 
+using ValueSet = std::unordered_set<Value*>;
+using ValueSetPtr = std::shared_ptr<std::unordered_set<Value*>>;
+
+Node* getLastUse(Value* v) {
+  auto last_use_node = v->node();
+  for (const auto& use : v->uses()) {
+    if (use.user->isAfter(last_use_node)) {
+      last_use_node = use.user;
+    }
+  }
+  return last_use_node;
+}
+
+void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
+  // This function first calculates aliasing sets,
+  // then calculates the last node each aliasing set is alive for.
+  // Then we go through each node, if it's a node which has an equivalent
+  // inplace node and the aliasing set for its input is dead afer this node, we
+  // inplace it. Then we merge the aliasing sets for the input and output of the
+  // node and extend the liveness of the set. To inplace a node you need to
+  // prove device and dtype of the input and output are the same, which we've
+  // already done, and prove that the output size is the same as the input size,
+  // which is achieved by explicit Broadcast nodes (which we inserted for other
+  // reasons).
+
+  auto aliasDb = torch::make_unique<AliasDb>(graph);
+  std::unordered_map<Value*, ValueSetPtr> alias_mapping;
+  ValueSet set;
+  ValueSetPtr input_set = std::make_shared<ValueSet>(set);
+  for (Value* v : graph->inputs()) {
+    if (v->type()->cast<TensorType>()) {
+      input_set->insert(v);
+      alias_mapping[v] = input_set;
+    }
+  }
+
+  auto merge_sets = [&](Value* existing, Value* new_v) {
+    if (alias_mapping[existing] == alias_mapping[new_v]) {
+      return;
+    }
+    auto existing_set = alias_mapping[existing];
+    auto set_to_remove = alias_mapping[new_v];
+    for (auto it = set_to_remove->begin(); it != set_to_remove->end(); it++) {
+      existing_set->insert(*it);
+      alias_mapping[*it] = existing_set;
+    }
+  };
+
+  for (Node* n : graph->nodes()) {
+    for (Value* output : n->outputs()) {
+      if (!output->type()->cast<TensorType>()) {
+        continue;
+      }
+
+      std::unordered_set<Value*> new_set = {output};
+      alias_mapping[output] = std::make_shared<ValueSet>(new_set);
+      for (Value* input : n->inputs()) {
+        if (aliasDb->mayAlias(input, output)) {
+          merge_sets(input, output);
+        }
+      }
+    }
+  }
+
+  std::unordered_map<ValueSetPtr, Node*> set_liveness;
```

Comment:
happy to separate them into different functions, but I'd have to thread around the different state that's computed in each function... I kind of thought it would be cleaner just to have one function that you read through linearly. somewhat related http://number-none.com/blow/john_carmack_on_inlined_code.html

Reply:
> http://number-none.com/blow/john_carmack_on_inlined_code.html

hehe, hard to argue w/ Carmack on anything. It's indeed hard to track a state if it's split across a bunch of functions. In this case, I was hoping that the state is pretty clearly delineated via `alias_mapping` and `liveness` and `node_to_inplace`, so they could become members fields in a liveness pass, but I agree it's not worth it since a possible readability improvement is highly subjective.

Reply:
I didnt separate out the `merge_sets` function, so i went part of the way :)